database/rrdcalc.c

```diff
@@ -499,274 +568,240 @@ inline RRDCALC *rrdcalc_create_from_template_unsafe(RRDHOST *host, RRDCALCTEMPLA
+    RRDHOST *host = ctr->rrdhost;
+
+    if(!rc->rrdset) {
+        if(ctr->from_rrdcalctemplate) {
+            rrdcontext_foreach_instance_with_rrdset_in_context(
+                host,
+                string2str(ctr->from_rrdcalctemplate->context),
+                rrdcalc_check_and_link_rrdset_callback,
+                rc);
+        }
+        else {
+            RRDSET *st;
+            rrdset_foreach_read(st, host) {
+                if(unlikely(rrdcalc_check_and_link_rrdset_callback(st, rc) == -1))
+                    break;
+            }
+            rrdset_foreach_done(st);
+        }
+    }
+}
 
-    // unlink it from RRDSET
-    if(rc->rrdset)
-        rrdcalc_unlink_from_rrdset_unsafe(rc);
+static void rrdcalc_rrdhost_delete_callback(const DICTIONARY_ITEM *item __maybe_unused, void *rrdcalc, void *rrdhost __maybe_unused) {
+    RRDCALC *rc = rrdcalc;
+    //RRDHOST *host = rrdhost;
 
-    // unlink it from RRDHOST
-    DOUBLE_LINKED_LIST_REMOVE_UNSAFE(host->host_alarms, rc, prev, next);
+    // TODO - unlink it
+    // TODO - free its variables
 
     rrdcalc_free(rc);
 }
 
-void rrdcalc_unlink_and_free_all_rrdset_alarms(RRDHOST *host, RRDSET *st) {
-    rrdhost_wrlock(host);
-    rrdset_wrlock(st);
-    while(st->alarms) {
-        RRDCALC *rc = st->alarms;
-
-        // unlink it from rddset
-        rrdcalc_unlink_from_rrdset_unsafe(rc);
-
-        // unlink it from rrdhost
-        DOUBLE_LINKED_LIST_REMOVE_UNSAFE(host->host_alarms, rc, prev, next);
+void rrdcalc_rrdhost_index_init(RRDHOST *host) {
+    if(!host->rrdcalc_root_index) {
+        host->rrdcalc_root_index = dictionary_create(DICTIONARY_FLAG_DONT_OVERWRITE_VALUE);
 
-        // free it
-        rrdcalc_free(rc);
+        dictionary_register_insert_callback(host->rrdcalc_root_index, rrdcalc_rrdhost_insert_callback, NULL);
+        dictionary_register_react_callback(host->rrdcalc_root_index, rrdcalc_rrdhost_react_callback, NULL);
+        dictionary_register_delete_callback(host->rrdcalc_root_index, rrdcalc_rrdhost_delete_callback, host);
     }
-    rrdset_unlock(st);
-    rrdhost_unlock(host);
 }
 
-void rrdcalc_unlink_and_free_all_rrdhost_alarms(RRDHOST *host) {
-    rrdhost_wrlock(host);
-    while(host->host_alarms) {
-        RRDCALC *rc = host->host_alarms;
-        RRDSET *st = rc->rrdset;
-
-        if(st) {
-            rrdset_wrlock(st);
-            rrdcalc_unlink_from_rrdset_unsafe(rc);
-            rrdset_unlock(st);
-        }
-
-        // unlink it from rrdhost
-        DOUBLE_LINKED_LIST_REMOVE_UNSAFE(host->host_alarms, rc, prev, next);
+void rrdcalc_rrdhost_index_destroy(RRDHOST *host) {
+    // destroy the id index last
+    dictionary_destroy(host->rrdcalc_root_index);
+    host->rrdcalc_root_index = NULL;
+}
 
-        // free it
-        rrdcalc_free(rc);
-    }
-    rrdhost_unlock(host);
+static size_t rrdcalc_key(char *dst, size_t dst_len, const char *chart, const char *alert) {
+    return snprintf(dst, dst_len, "%s/%s", chart, alert);
```

Comment:
Is there a special reason to use `snprintf` instead `snprintfz`?

Reply:
thanks! fixed it!